logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,388 @@
+            final SourceWithMetadata source = v.getSourceWithMetadata();
+            nodes.add(pipeline.buildInput(
+                RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+                RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+            ));
+        });
+        return nodes;
+    }
+
+    /**
+     * Converts plugin arguments from the format provided by {@link PipelineIR} into coercible
+     * Ruby types.
+     * @param def PluginDefinition as provided by {@link PipelineIR}
+     * @return RubyHash of plugin arguments as understood by {@link RubyIntegration.Pipeline}
+     * methods
+     */
+    private RubyHash convertArgs(final PluginDefinition def) {
+        final RubyHash converted = RubyHash.newHash(RubyUtil.RUBY);
+        for (final Map.Entry<String, Object> entry : def.getArguments().entrySet()) {
+            final Object value = entry.getValue();
+            final String key = entry.getKey();
+            final Object toput;
+            if (value instanceof PluginStatement) {
+                final PluginDefinition codec = ((PluginStatement) value).getPluginDefinition();
+                toput = pipeline.buildCodec(
+                    RubyUtil.RUBY.newString(codec.getName()),
+                    Rubyfier.deep(RubyUtil.RUBY, codec.getArguments())
+                );
+            } else {
+                toput = value;
+            }
+            converted.put(key, toput);
+        }
+        return converted;
+    }
+
+    /**
+     * Compiles a {@link RubyIntegration.Filter} from a given {@link PluginVertex}.
+     * @param vertex Filter {@link PluginVertex}
+     * @return Compiled {@link RubyIntegration.Filter}
+     */
+    private RubyIntegration.Filter buildFilter(final PluginVertex vertex) {
+        final PluginDefinition def = vertex.getPluginDefinition();
+        final SourceWithMetadata source = vertex.getSourceWithMetadata();
+        return pipeline.buildFilter(
+            RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+            RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+        );
+    }
+
+    /**
+     * Checks if a certain {@link Vertex} represents a {@link RubyIntegration.Filter}.
+     * @param vertex Vertex to check
+     * @return True iff {@link Vertex} represents a {@link RubyIntegration.Filter}
+     */
+    private boolean isFilter(final Vertex vertex) {
+        return filters.containsKey(vertex.getId());
+    }
+
+    /**
+     * Checks if a certain {@link Vertex} represents a {@link RubyIntegration.Output}.
+     * @param vertex Vertex to check
+     * @return True iff {@link Vertex} represents a {@link RubyIntegration.Output}
+     */
+    private boolean isOutput(final Vertex vertex) {
+        return outputs.containsKey(vertex.getId());
+    }
+
+    /**
+     * Compiles the next level of the execution from the given {@link Vertex} or simply return
```

Comment:
I'm also wondering if passing `parents` around makes the code harder to read. It might be simpler if we just used the map of `Vertex->Dataset` with the `ROOT_DATASETS` as a default value.

Reply:
@andrewvc yea you're right. The design here is somewhat evil in hindsight ... I think the problem starts here:

```java
    public Dataset buildExecution() {
        iffs.clear();
        plugins.clear();
```

I'm reusing some collections and passing down the `parents` ... in actuality, there should probably not be any mutable parts here ... I'll improve the design :)